[compiler-rt][windows] Test fixups for MSVC.

barcharcraz opened the pull request:
- add XFAIL/UNSUPPORTED annotations for tests run wtih real MSVC 
- macroify usages of clang-specific attributes in asan tests

These are most of the changes that are required to get things running with MSVC, however there are some remaining build-flag tweaks

Commit:
Test fixups for MSVC.
add XFAIL/UNSUPPORTED annotations for tests run wtih real MSVC
macroify usages of clang-specific attributes in asan tests

mstorsjo commented on compiler-rt/test/asan/TestCases/Windows/dll_intercept_memset.cpp:
```diff
@@ -1,9 +1,9 @@
-// RUN: %clang_cl_asan %Od %p/dll_host.cpp %Fe%t
-// RUN: %clang_cl_asan -Wno-fortify-source %LD %Od %s %Fe%t.dll
+// RUN: %clang_cl_asan %Od /Oy- %p/dll_host.cpp %Fe%t
```
Don't use an unconditional `/Oy-` here - we run these tests with mingw compilers as well, with arguments replaced with corresponding ones for the GNU style compiler driver - hence substitutions like `%Od` and similar here; see `test/asan/lit.cfg.py`.

barcharcraz replied:
I think I actually have the fixes for this in a follow-up.

I'll append that commit here and we can merge both at once if you'd like.

I could alternatively create another (stacked) PR with those changes

mstorsjo replied:
Hmm, depending on how big that extra change is, it may be ok to add that change here. But this change is in itself already quite big, so it could also be clearer to just back out this change from this particular PR (so it doesn't fix everything for MSVC, just the bulk of things) and then do that change in a separate step afterwards, for the sake of easier readability/reviewability.

barcharcraz replied:
Turns out we hadn't fixed this internally, but I've added it here.

I decided to just add the substitution, although using `%if` could also make sense considering this is the only place where /Oy- shows up.

mstorsjo commented on compiler-rt/test/asan/TestCases/Windows/dll_operator_array_new_with_dtor_left_oob.cpp:
```diff
@@ -2,12 +2,18 @@
 
-int __attribute__((noinline, optnone)) hide(int x) { return x; }
+int ATTRIBUTE_NOINLINE
+#if __clang__ || !defined(_MSC_VER)
```
I guess this should be `defined(__clang__)` instead of plain `__clang__`?

barcharcraz replied:
I think this is somewhat stylistic. I can switch to defined instead of relying on the substitution of undefined identifiers for 0 though, it's a reasonable style.

mstorsjo replied:
Yeah I think we in general would prefer to use `defined()` rather than relying on expansion to zero (although I'm not sure if we have a written down policy about it), so the code would be clean for `-Wundef` too.

mstorsjo commented on compiler-rt/test/asan/TestCases/Windows/issue64990.cpp:
```diff
@@ -1,9 +1,13 @@
 
 // UNSUPPORTED: target={{.*-windows-gnu}}
 
+#if _MSC_VER && !defined(__clang__)
```
`defined(_MSC_VER)`

mstorsjo commented on compiler-rt/test/asan/TestCases/Windows/issue64990.cpp:
```diff
@@ -12,7 +16,11 @@ int main(int argc, char **argv) {
     throw 1;
   } catch (...) {
     // Make asan generate call to __asan_memcpy inside the EH pad.
+#if _MSC_VER && !defined(__clang__)
```
Ditto

mstorsjo commented on compiler-rt/test/asan/TestCases/Windows/null_deref.cpp:
```diff
@@ -1,15 +0,0 @@
-  ptr[10]++;  // BOOM
-}
-int main() {
-  NullDeref((int*)0);
```
What's up with this deleted test?

barcharcraz replied:
this contained the following comment:
> // FIXME: merge this with the common null_deref test when we can run common
// tests on Windows.

and you'll note that TestCases/null_deref.cpp (the mentioned common test) was also modified to work on windows.

So the deletion is just fixing this fixme :D

mstorsjo replied:
Right - but I don't see any `XFAIL` being removed from `TestCases/null_deref.cpp`, so to me it seems like that test was already being executed on Windows, with clang - this change just makes it work with MSVC too?

(I.e. in other word, this FIXME that you're removing was already ripe to be removed before this patch.)

barcharcraz replied:
yep, it was indeed ripe to be removed prior to this patch.

If you like I could extract this change into its own PR I suppose

vitalybuka replied:
If we started over, this PR could be split into set of PRs where each PR contains same patter of changes. Easier to review.

mstorsjo commented on compiler-rt/test/asan/TestCases/alloca_constant_size.cpp:
```diff
@@ -7,6 +7,9 @@
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
+#if _MSC_VER
```
`defined(_MSC_VER)`

mstorsjo commented on compiler-rt/test/asan/TestCases/Windows/null_deref_multiple_dlls.cpp:
```diff
@@ -27,8 +28,8 @@ extern "C" {
 __declspec(dllexport) void foo1() {}
 }
 #elif defined(DLL2)
-extern "C" {
```
Here you lost an `extern "C" {`, and end up with unbalanced parentheses.

mstorsjo commented on compiler-rt/test/asan/TestCases/interception_failure_test.cpp:
```diff
@@ -18,6 +18,10 @@
 // it works with the dynamic runtime.
 // XFAIL: target={{.*netbsd.*}} && !asan-dynamic-runtime
 
+#if _MSC_VER && !defined(__clang__)
```
`defined(_MSC_VER)` (Also please grep for more cases of this in the patch, I probably won't comment on all of them.)

barcharcraz replied:
done and I've also grepped for relevant cases, I avoided changing anything outside of compiler-rt

mstorsjo commented on compiler-rt/test/asan/TestCases/throw_invoke_test.cpp:
```diff
@@ -34,7 +34,7 @@ void Throw() {
 }
 
-__attribute__((noinline))
+ATTRIBUTE_NOINLINE
```
This is missing an include of `defines.h`

mstorsjo commented on compiler-rt/test/asan/TestCases/use-after-scope-loop-bug.cpp:
```diff
@@ -1,4 +1,4 @@
-// RUN: %clangxx_asan -O1 %s -o %t && not %run %t 2>&1 | FileCheck %s
+// RUN: %clangxx_asan %if MSVC %{ /Od %} %else %{ /O1 %} %s -o %t && not %run %t 2>&1 | FileCheck %s
```
This was using `-O1` which works both with MSVC, clang-cl and the clang driver, don't switch to `/O1` which isn't accepted by the GNU style clang driver.

mstorsjo reviewed (commented):
I did a cursory review and pointed out a couple of issues I found, but I'd appreciate @vitalybuka's opinions too.

Regarding keeping these tests working in mingw configurations, I set up a github actions workflow that you can try out to run these tests - see the second commit at https://github.com/mstorsjo/llvm-project/commits/gha-mingw-compiler-rt. I tested your PR on top of it, with the following results: https://github.com/mstorsjo/llvm-project/actions/runs/11029615284/job/30632424312

(Ideally we'd have things like this integrated into regular precommit CI testing. I haven't tested running tests with clang-cl in similar configurations on github actions though.)

Finally, a small nit - the PR subject could use some tag to indicate that this is related to compiler-rt, to make it clearer once looking at the final commit as part of the whole commit log.

compnerd commented on compiler-rt/test/asan/TestCases/Windows/operator_array_new_with_dtor_left_oob.cpp:
```diff
@@ -1,12 +1,18 @@
+#else
+__attribute__((optnone))
+#endif
+int ATTRIBUTE_NOINLINE hide(int x) { return x; }
```
You need too restore the optimization level around this with MSVC.

compnerd commented on compiler-rt/test/asan/TestCases/ill.cpp:
```diff
@@ -2,12 +2,16 @@
 
+#if _MSC_VER && !defined(__clang__)
+#define __builtin_trap() __asm ud2;
+#endif
```
This isn't safe - the `__asm` is limited to x86 and not available at x86_64.

barcharcraz committed:
Add substitution for /Oy-/-fno-omit-frame-pointer

This makes the dll_intercept_memset test work with mingw

barcharcraz committed:
Use #if defined(...) instead of #if ...

barcharcraz committed:
Fix a few mangled test files

These were mangled in the process of pulling them out of the msvc fork

- null_deref_multiple_dlls.cpp -- extern "C" { got removed
- throw_invoke_test.cpp -- #include "defines.h" is needed for
  ATTRIBUTE_NOINLINE, this test is unsupported on msvc so this was not
  caught

barcharcraz committed:
use -O1 instead of /O1 in use-after-scope-loop-bug

This is compatible with the "real" gnu toolchain

barcharcraz committed:
add a comment clarifying arch support in ill.cpp

With real MSVC this test only supports i386

barcharcraz committed:
operator_array_new_with_dtor_left_oob: turn opt back on

unlike __attribute__((optnone)) MSVC's #pragma optimize("", off)
needs to be turned back on

barcharcraz committed:
clang-format everything

mstorsjo commented on compiler-rt/test/asan/TestCases/ill.cpp:
```diff
@@ -2,12 +2,18 @@
 #endif
 
+// note: test is limited to i386 only ("asan-32-bits") when using "real" MSVC
+//  see the  REQUIRES: clause above
```
This `REQUIRES:` string in the comment here ends up intercepted by `lit`, and can't be parsed, so the test ends up in the category Unresolved, and not executed at all. I guess you at the very least need to drop the `:`, possibly differentiate it further.

mstorsjo commented on compiler-rt/test/asan/TestCases/Windows/dll_intercept_memset.cpp:
```diff
@@ -1,9 +1,9 @@
-// RUN: %clang_cl_asan %Od %p/dll_host.cpp %Fe%t
-// RUN: %clang_cl_asan -Wno-fortify-source %LD %Od %s %Fe%t.dll
+// RUN: %clang_cl_asan %Od %Oy- %p/dll_host.cpp %Fe%t
+// RUN: %clang_cl_asan %if !MSVC %{ -Wno-fortify-source %} /Oy- %LD %Od %s %Fe%t.dll
```
This line still has got a literal `/Oy-`, not the substituted form.

vitalybuka commented on compiler-rt/test/asan/TestCases/Windows/coverage-basic.cpp:
```diff
@@ -4,6 +4,7 @@
 // RUN: %env_asan_opts=coverage=1 %run ./test.exe
 //
 // RUN: %sancov print *.sancov | FileCheck %s
+// XFAIL: MSVC
```
for consistency: MSVC -> msvc

vitalybuka replied:
"or opposite" but consistent.

vitalybuka commented on compiler-rt/test/asan/TestCases/alloca_instruments_all_paddings.cpp:
```diff
@@ -2,6 +2,7 @@
 // RUN: %clangxx_asan -O3 -mllvm -asan-instrument-dynamic-allocas %s -o %t
 // RUN: %run %t 2>&1
 //
+// UNSUPPORTED: MSVC
```
It would be nice to comment each XFAIL and UNSUPPORTED with a reason.

vitalybuka commented on compiler-rt/test/asan/TestCases/debug_locate.cpp:
```diff
@@ -1,3 +1,4 @@
+// UNSUPPORTED: MSVC
```
unusual place for UNSUPPORTED

vitalybuka commented on compiler-rt/test/asan/TestCases/defines.h:
```diff
@@ -0,0 +1,31 @@
+#pragma once
```
we usually use ifdef kind of guard

vitalybuka commented on compiler-rt/test/asan/TestCases/inline.cpp:
```diff
@@ -2,10 +2,14 @@
 // be normally inlined.
+//
+// XFAIL: MSVC
+// ^ MSVC has this bug
```
which bug?

barcharcraz replied:
if the function is inlined the no_sanitize_ address equivalent in msvc (`__declspec(no_sanitize_address)`) does not apply

vitalybuka reviewed (approved):
LGTM with comments

set of PRs could be nicer

barcharcraz commented:
> LGTM with comments
> 
> set of PRs could be nicer

By "set of PRs" do you mean splitting this into multiple PRs or posting the follow-ups immediately?

barcharcraz committed:
Use an include guard instead of #pragma once

In asan/TestCases/defines.h

barcharcraz committed:
Move UNSUPPORTED line to below opening comment

barcharcraz committed:
MSVC -> msvc

barcharcraz committed:
say why things are unsupported on MSVC

barcharcraz committed:
reword comment in asan/TestCases/inline.cpp

Commit:
clang-format comment

vitalybuka commented:
> > LGTM with comments
> > set of PRs could be nicer
> 
> By "set of PRs" do you mean splitting this into multiple PRs or posting the follow-ups immediately?

Both: Splitting, and landing, and followups in new PRs

And make sure to "re-request" review if you expect reviewer attention.
